// grammy/out/platform.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InputFile = exports.baseFetchConfig = exports.itrToStream = exports.streamFile = exports.debug = void 0;
const https_1 = require("https");
const path_1 = require("path");
const stream_1 = require("stream");
__exportStar(require("@grammyjs/types"), exports);
var debug_1 = require("debug");
Object.defineProperty(exports, "debug", { enumerable: true, get: function () { return debug_1.debug; } });
// Turn a file path into an AsyncIterable<Uint8Array>
var fs_1 = require("fs");
Object.defineProperty(exports, "streamFile", { enumerable: true, get: function () { return fs_1.createReadStream; } });
// Turn an AsyncIterable<Uint8Array> into a stream
const itrToStream = (itr) => stream_1.Readable.from(itr, { objectMode: false });
exports.itrToStream = itrToStream;
// Base configuration for `fetch` calls
exports.baseFetchConfig = {
    compress: true,
    agent: new https_1.Agent({ keepAlive: true }),
};
/** This object represents the contents of a file to be uploaded. Must be posted using multipart/form-data in the usual way that files are uploaded via the browser. */
class InputFile {
    /**
     * Constructs an `InputFile` that can be used in the API to send files.
     *
     * @param file A path to a local file or a `Buffer` or a `fs.ReadStream` that specifies the file data
     * @param filename Optional name of the file
     */
    constructor(file, filename) {
        this.file = file;
        if (filename === undefined && typeof file === "string")
            filename = path_1.basename(file);
        this.filename = filename;
    }
}
exports.InputFile = InputFile;
